ENH: sparse: Add indexing for 1D arrays

dschult opened the pull request:
Adds support for indexing with 1D arrays.

Builds on #19833 (csr-1d) to provide indexing for 1D CSR arrays, though this code also works for DOK with minor changes after #19715 (dok-1d) gets merged.

The diff will be much easier to read after #19833 is merged. Look at only the last commit to see the diff for just this PR. I will rebase as needed, but I think the CSR PR will be merged soon and the diff here will become cleaner.

- adds `test_indexing1d.py`
- refactors `IndexMixin._validate_indices` to use an ndim-independent approach for both 1d and 2d.
- the supported formats will need methods `_get_int`, `_get_slice` and `_get_array`. These are dispatched to after processing the index.
- this tries to incorporate all improvements from the recently merged #19957 (and doesn't change those tests). 
- the ndim-independent approach means some helper functions are no longer needed and/or have been inlined. The two remaining helper functions are now methods so they have access to `self.ndim`. They are `_as_indices` and `_compatible_boolean_array`.

dschult committed:
add csr 1d and test_arithmetic1d

dschult committed:
add csr 1d and test_arithmetic1d

dschult committed:
clean up after rebase of _mul to _matmul namechange

dschult committed:
clean up after rebase of _mul to _matmul namechange

dschult committed:
change matmul to produce 0d with vec @ vec. Adjust tests. clean up comments

dschult committed:
change matmul to produce 0d with vec @ vec. Adjust tests. clean up comments

dschult committed:
update _index.py for 1d, refactor _validate_indices, add tests.

Move TestGetSet1d to indexing1d.py. Simplify helper functions.

dschult committed:
Add tests for 1d@1d. Fix 1d on the right of @. Slow to convert to csc

dschult committed:
Add tests for 1d@1d. Fix 1d on the right of @. Slow to convert to csc

j-bowhay commented on scipy/sparse/tests/test_arithmetic1d.py:
```diff
@@ -0,0 +1,337 @@
+            assert isinstance(a @ a.tocoo(), np.ndarray)
+            for m in [a, b, c]:
+                assert m @ m == a.toarray() @ a.toarray()
+                assert np.array_equal(m.dtype, mytype)
```
style nit but here and elsewhere why not use `assert_equal` / `assert_allclose`?

dschult replied:
Somehow I got the message that the preference style choice with `pytest` is to use `assert` and then an expression. The "old-style" node api encouraged testing functions like `assert_equal`. With `pytest` stays closer to builtin python commands. It's also nice that `array_equal` is in the main `numpy` namespace so no additional imports are needed. But if SciPy has a style for testing that prefers the numpy testing functions, I can change it easily. Let me know.

dschult committed:
update _index.py: None indexes, validate returns index and new_shape, fix fancy

This should set us up for nD indexing when the time comes.
Note: ndindex could work for much of this -- but not for sparse array boolean
In the future we could maybe implement sparse integer array indexing.

We should think about whether 0D return should be sparse array or ndarray.
Currently ndarray.

This doesn't make reduction functions return sparse arrays yet -- only indexing

dschult committed:
fixup _coo_to_compressed from setdiag fix to work for 1d

tylerjereddy commented on scipy/sparse/tests/test_arithmetic1d.py:
```diff
@@ -0,0 +1,337 @@
+    def test_imag(self, spcreator):
+        D = np.array([1 + 3j, 2 - 4j])
+        A = spcreator(D)
+        assert np.array_equal(A.imag.toarray(), D.imag)
```
for these and other similar cases I'd say `numpy.testing` functions are the right way to go yeah.

dschult committed:
convert assert tests to np.testing assert_equal and assert_allclose

dschult committed:
lint fixes

dschult committed:
lint fixes

dschult committed:
fix error msg issue due to lint fix

dschult commented:
I've changed the tests to use np.testing `assert_equal` and `assert_allclose` in these new tests. I'll create a separate PR to change the new-but-merged-tests to use them too.

I've also added indexing support for np.newaxis/None (though it can't make 3D+). And that was an excuse to revamp some of the getitem/setitem code to be easier to move to nd.  The ndindex library gives almost everything we need, but doesn't support boolean sparse arrays as indexes. So we can now index 2d and 1d sparse arrays.

I believe the test failures are not related to this PR.

Note: This PR does not change reduction operations to have them return 1D sparse arrays. That will be a separate PR.

Note: Another separate PR: We need to decide how to handle 0D. `A[3,4]` should return a 0D object as per array_api. But should it return a numpy or scipy.sparse 0D object? In general, an array type "should" return it's own type when indexed. But sparse is not a standard array type -- we don't expect the entire array api to be implemented. But the answer to this question will impact this same code.

dschult committed:
add DOK support for 1D indexing.

dschult committed:
fix _coo.py review comments; add copy flag to 1D tocsr

dschult committed:
fix other review comments

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -32,74 +26,122 @@ class IndexMixin:
+
+        else:  # 2D array
+            ixtypes = tuple(
+                0 if isinstance(ix, INT_TYPES) else 1 if isinstance(ix, slice) else 2
```
```
SCALAR = 0
SLICE = 1
ARRAY = 2
```

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -32,74 +26,122 @@ class IndexMixin:
+                res = self._get_array(idx)
+
+        else:  # 2D array
+            ixtypes = tuple(
```
Let's move this up before the 1d case

dschult committed:
rewrite fin new_shape logic. make copy of coords if asked.

dschult committed:
revert dispatch logic to 1D and 1D explicit treatment

dschult committed:
Merge branch 'csr-1d' into index-1d

dschult committed:
Merge branch 'main' into index-1d

dschult committed:
Adjust after merge

dschult committed:
Merge branch 'index-1d' of https://github.com/dschult/scipy into index-1d

dschult committed:
linting

dschult committed:
fix test

dschult committed:
try handling Nones with scalar

dschult committed:
reorder getitem/setitem

dschult committed:
Merge branch 'main' into index-1d

dschult committed:
remove override of get/setitem in csr

dschult committed:
handle style changes

dschult committed:
update _get_int and friends in compressed

perimosocordiae committed:
Merge branch 'main' into index-1d

perimosocordiae commented on scipy/sparse/_csr.py:
```diff
@@ -223,6 +189,38 @@ def _getcol(self, i):
+        col = np.asarray(idx, dtype=idx_dtype)
+        val = np.empty(row.size, dtype=self.dtype)
+        csr_sample_values(M, N, self.indptr, self.indices, self.data,
+                          row.size, row.ravel(), col.ravel(), val)
```
No need to .ravel() here, as we know these are already 1d.

perimosocordiae commented on scipy/sparse/_csr.py:
```diff
@@ -276,6 +274,14 @@ def _get_arrayXslice(self, row, col):
         return self._major_index_fancy(row)._get_submatrix(minor=col)
 
+    def _set_int(self, idx, x):
+        major, minor = self._swap((0, idx))
```
Since we only support 1d for CSR format, this can be simplified to `self._set_many(0, idx, x)`

perimosocordiae commented on scipy/sparse/_dok.py:
```diff
@@ -140,27 +140,35 @@ def get(self, key, default=0.0):
+
+    def _get_array(self, idx):
+        idx = np.asarray(idx)
+        if len(idx.shape) == 0:
```
if idx.ndim == 0

perimosocordiae commented on scipy/sparse/_dok.py:
```diff
@@ -140,27 +140,35 @@ def get(self, key, default=0.0):
+            else:
+                new_idx = np.unravel_index(np.arange(len(dok_vals)), idx.shape)
+                new_idx = new_idx[0] if len(new_idx) == 1 else zip(*new_idx)
+                # zip could use 'strict=True' With Python 3.10+
```
We now require 3.10+, so go for it!

perimosocordiae commented on scipy/sparse/_dok.py:
```diff
@@ -232,30 +240,27 @@ def _get_arrayXarray(self, row, col):
+        x_set = x.ravel()
+        if len(idx_set) != len(x_set):
+            if len(x_set) == 1:
+                x_set = np.array([x_set[0]] * len(idx_set), dtype=self.dtype)
```
`np.full(len(idx_set), x_set[0], dtype=self.dtype)`

perimosocordiae committed:
Update test_common1d.py

dschult committed:
adopt review suggestions

dschult commented:
I've made the suggested changes. Thanks for the note about Python 3.10+ being supported (3.9 dropped)

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -32,74 +26,125 @@ class IndexMixin:
+                res = self._get_arrayXslice(row, col)
+            # arrayXarray preprocess
+            elif row.ndim == 2 and row.shape[1] == 1\
+                and (col.ndim == 1 or col.shape[0] == 1):
```
You can avoid the backslash by wrapping the whole condition in parens:

```python
elif (row.ndim == 2 and row.shape[1] == 1
      and (col.ndim == 1 or col.shape[0] == 1)):
```

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -32,74 +26,125 @@ class IndexMixin:
+                    res = self._get_arrayXarray(row, col)
+
+        # package the result and return
+        if isinstance(self, sparray):
```
```python
if isinstance(self, sparray) and res.shape != new_shape:
```

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -32,74 +26,125 @@ class IndexMixin:
+
+            if issparse(x):
+                x = x.toarray()
+            x = np.asarray(x, dtype=self.dtype)
```
put this in an `else` block?

dschult replied:
We need to check/change the dtype even if `x` was sparse to begin with. So I think we should not have the `else` block, but I think the `asarray` is confusing and inefficient here and maybe why the else is in question. We are only changing the dtype of the array in line 120. So I think we should use `x = x.astype(dtype=self.dtype, copy=False)`

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -32,74 +26,125 @@ class IndexMixin:
+                x = np.broadcast_to(x, idx.shape)
+            if x.size == 0:
+                return
+            self._set_array(idx, x)
```
```python
if x.size != 0:
    self._set_array(idx, x)
return
```

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -151,52 +195,85 @@ def __setitem__(self, key, x):
-            axis_name: str
-        ) -> npt.NDArray[np.int_]:
-            if len(idx) != axis_size:
+        """Returns index tuple of intended result"""
```
Mention that we now return two things: the index tuple as well as the resulting shape.

perimosocordiae commented on scipy/sparse/_index.py:
```diff
@@ -151,52 +195,85 @@ def __setitem__(self, key, x):
+        if array_indices:
+            idx_shape = list(index[array_indices[0]].shape) + idx_shape
+        if len(idx_shape) > 2:
+            ndim = len(idx_shape)
```
```python

if (ndim := len(idx_shape)) > 2:
    raise IndexError(...)
```

perimosocordiae commented on scipy/sparse/tests/test_indexing1d.py:
```diff
@@ -0,0 +1,504 @@
+@contextlib.contextmanager
+def check_remains_sorted(X):
+    """Checks that sorted indices property is retained through an operation"""
+    if not hasattr(X, 'has_sorted_indices') or not X.has_sorted_indices:
```
```python
yield
if not getattr(X, 'has_sorted_indices', False):
    return
```

dschult committed:
implement suggestions

dschult commented:
I updated as suggested (and realized why the else block suggestion was correct -- we don't need to convert the dtype at that location.

This should be ready. The failing test seems unrelated.

@stefanv you looked at this some a while back -- can you take another look?

perimosocordiae commented:
This looks ready to merge from my end. @stefanv feel free to push the button if you agree.

dschult commented:
Thoughts for release notes: 
In `New features` section for sparse:

- sparse arrays now support indexing for 1D-arrays using ints, slices and arrays.
  Methods `min`, `max`, `nanmin`, `nanmax` return 1D COO arrays when ``axis``
  is specified. 

Getting ahead of myself, but feedback is welcome:
Once the migration guide document is available we can make a `Highlights` entry something like:

- sparse arrays are now fully functional replacements for sparse matrices.
  Consider converting your code to use sparse arrays. See `migration_to_sparray.rst`
  for guidelines to this process and open an issue if you have questions. We have
  dedicated reviewer bandwidth to help with these conversion over the next two releases.

stefanv commented on scipy/sparse/_csr.py:
```diff
@@ -223,6 +189,38 @@ def _getcol(self, i):
+            if spot.size:
+                return self.data[spot[0]]
+            return self.data.dtype.type(0)
+        raise IndexError(f'index ({idx}) out of range')
```
It's a bit strange to have the error removed from the check so far; it's not necessary to change for this PR, but I think it'd be more intuitive to write:

```
if (idx < 0) or (idx >= self.shape[0]):
    raise IndexError(...)
```

Is the current if check correct? It has `<=` in both positions, but presumably `idx` should be strictly less than `self.shape[0]`.
```

What about negative indices?

dschult replied:
I agree with your style comment -- and I agree that the `<=` should be `< self.shape[0]`.

Negative indices are handled in the `_validate_indices` method. 

Which reminded me that `_validate_indices` also checks the size of `idx`, so there is no need to check it here. That is why the incorrect `<=` check wasn't flagged by the existing test in `test_indexing1d.py::test_getelement` for index values just beyond the limits of `shape`. 

Thanks for flagging this -- I have removed this "if/raise" statement as it is redundant with logic in `_validate_index`.

stefanv commented on scipy/sparse/_dok.py:
```diff
@@ -140,27 +140,34 @@ def get(self, key, default=0.0):
+        return self._get_array(list(i_range))
+
+    def _get_array(self, idx):
+        idx = np.asarray(idx)
```
This call can blow up the range, is that something we are worried about?

dschult replied:
This is how the method `_asindices` handles the input `idx`. I can guess that we aren't worried about it blowing up because we don't support sparse non-boolean arrays as indices, and we handle boolean arrays (in `_validate_indices`) by converting them using `ix.nonzero()`.

stefanv commented on scipy/sparse/_index.py:
```diff
@@ -32,74 +26,124 @@ class IndexMixin:
+            # package the result and return
+            if not isinstance(self, sparray):
+                return res
+            if res.shape == () and new_shape != ():
+                if len(new_shape) == 1:
+                    return self.__class__([res], shape=new_shape, dtype=self.dtype)
+                if len(new_shape) == 2:
+                    return self.__class__([[res]], shape=new_shape, dtype=self.dtype)
```
This is probably fine, but I don't 100% get which case we're handling here.

dschult replied:
I've added a comment indicating that this is to handle multiple `np.newaxis` indices, i.e. `A[3, 4, None, None]`.

stefanv commented on scipy/sparse/_index.py:
```diff
@@ -32,74 +26,124 @@ class IndexMixin:
+                N = len(idx_range)
+                if N == 1 and x.size == 1:
+                    self._set_int(idx_range[0], x.flat[0])
+                idx = np.arange(*idx.indices(self.shape[0]))
```
This does densify the range above; is that OK?

dschult replied:
I was basing this idiom on the 2D case from this module, and `get_arrayXslice` in _csr.py.  I think the comment above isn't clear. We use a Python `range` to check for the special case when the slice has a single element -- and we call `_set_int` instead of `_set_array`.  Part of the lack of clarity is that this shortcut should be finished with a `return` and it isn't. So, in fact the code in the PR was setting the element once in `_set_int` and then again in `_set_array` later on.  That's my mistake. And the tests don't catch it because it is just doing extra work -- not creating a wrong answer.

I have added a comment and a return statement to make this more clear, and actually take the shortcut when we can.

stefanv commented on scipy/sparse/_index.py:
```diff
@@ -293,100 +383,23 @@ def _set_arrayXarray_sparse(self, row, col, x):
+def _compatible_boolean_index(idx, desired_ndim):
+    """Check for boolean array or array-like. peek before asarray for array-like"""
+    # assume already an array if attr ndim exists: skip to bottom
+    if not hasattr(idx, 'ndim'):
```
This feels a bit risky; but not sure there's any single attribute that would tell us whether we are dealing with a numpy array or any array-like derivative. Maybe `.__array__`, but again not sure what external arrays would implement; we'd perhaps have to check with someone who works on the Array API.

dschult replied:
I agree that it feels risky -- but it is following the logic of the previous code (the removed line 384 below) which used `ndim` as an indicator of a compatible array type.

I will look for a more appropriate check for compatible array types, but I think that should be a separate PR.
I have added/improved comments explaining that this assumption is being made. Also for better clarity, I added `return None` at the end of this function to be more clear that returning None carries information here .

dschult committed:
address review

remove superfluous check of idx
add multple comments to clarify logic
add clarifying `return None` at end of _compatible_boolean_index

stefanv commented:
Thank you, @dschult!
